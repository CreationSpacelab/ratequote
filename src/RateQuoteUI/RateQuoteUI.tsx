import React from 'react';


const RateQuoteUI: React.FC = () => {
  
  return (
    <>
      Hi
    </>
  );
}

export default RateQuoteUI;


