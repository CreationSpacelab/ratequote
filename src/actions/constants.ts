export const ADD_QUOTES = "ADD_QUOTES";
export const START_LOADING_DATA = "START_LOADING_DATA";
export const END_LOADING_DATA = "END_LOADING_DATA";
export const FETCH_ERROR = "FETCH_ERROR";
