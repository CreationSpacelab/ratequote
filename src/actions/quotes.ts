import { ThunkDispatch } from "redux-thunk";

import { QuoteType, RateQuoteUIProps } from "../../src/quoteTypes";
import config from "../config";

import {
  ADD_QUOTES,
  FETCH_ERROR,
  START_LOADING_DATA,
  END_LOADING_DATA,
} from "./constants";

export const initiateAddQuotes: (
  params: RateQuoteUIProps
) => (
  dispatch: ThunkDispatch<QuoteType, void, Action>
) => Promise<QuoteType> = (params: RateQuoteUIProps) => {
  // Create a params for fetch url
  const fetchParams = Object.entries(params)
    .map(([k, v]) => {
      // Convert the loan size value to unformatted string
      if (k === "loanSize") return `${k}=${v.split(",").join("")}`;
      return `${k}=${v}`;
    })
    .join("&");

  // Add fetch headers and fetch method
  const fetchConfig = {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      Authorization: config.API_KEY,
    },
  };

  return async (dispatch: ThunkDispatch<QuoteType, void, Action>) => {
    /* To improve user engagement, loader is active while fetching data. Once the data is retrieved,
      the loader will change to inactive.
    */
    try {
      dispatch(startLoading());
      await fetch(
        `https://ss6b2ke2ca.execute-api.us-east-1.amazonaws.com/Prod/quotes?${fetchParams}`,
        fetchConfig
      )
        .then((response: Response) => response.json())
        .then((res: { rateQuotes: QuoteType[] }) => {
          // Dispatch the quotes to the redux store
          return dispatch(addQuotes([res.rateQuotes]));
        });
    } catch (error) {
      dispatch(fetchError(error));
      return error;
    }
    // Once data is retrieved end loading
    setTimeout(() => {
      dispatch(endLoading());
    }, 2000);
  };
};

// Setting the type of any object and hint to the compiler about immutability.
export const addQuotes = (quotes: QuoteType[][]) => {
  return <const>{
    type: ADD_QUOTES,
    quotes,
  };
};

export const fetchError = (e: string) => {
  return <const>{
    type: FETCH_ERROR,
    error: e,
  };
};

export const startLoading = () => {
  return <const>{
    type: START_LOADING_DATA,
    loading: true,
  };
};

export const endLoading = () => {
  return <const>{
    type: END_LOADING_DATA,
    loading: false,
  };
};

export type Action = ReturnType<
  typeof addQuotes | typeof fetchError | typeof startLoading | typeof endLoading
>;
