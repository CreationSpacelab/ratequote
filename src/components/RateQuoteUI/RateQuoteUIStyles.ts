import styled from "styled-components";

import { theme } from "../core/colors";

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 600px;
  margin: 32px auto;
  padding: 32px;
`;

export const Container = styled.div`
  border: 1px solid ${theme.lightgray};
  border-radius: 4px;
  width: fit-content;
  margin: 0 auto;
  min-width: 780px;
  max-width: 1024px;
  position: relative;
  width: 100%;
`;

export const Loader = styled.div`
  position: absolute;
  bottom: -50px;
  left: 50%;
  transform: translateX(-50%);

  &,
  &:before,
  &:after {
    border-radius: 50%;
    width: 30px;
    height: 30px;
    animation: loaderframe 1.4s infinite cubic-bezier(0.39, 0.575, 0.565, 1)
      both;
  }

  & {
    color: ${theme.darkgray};
    font-size: 10px;
    text-indent: -159984px;
    transform: translateZ(0);
    animation-delay: -0.16s;
  }

  &:before,
  &:after {
    content: "";
    position: absolute;
    top: 0;
  }
  &:before {
    left: -3.5em;
    animation-delay: -0.32s;
  }
  &:after {
    left: 3.5em;
  }

  @keyframes loaderframe {
    0%,
    80%,
    100% {
      box-shadow: 0 30px 0 -20px;
    }
    40% {
      box-shadow: 0 30px 0 0;
    }
  }
`;
