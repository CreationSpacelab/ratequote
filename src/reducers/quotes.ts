import {
  ADD_QUOTES,
  START_LOADING_DATA,
  END_LOADING_DATA,
  FETCH_ERROR
} from "../actions/constants";
import { QuoteReducerProps } from "../quoteTypes";

const initialState: Partial<QuoteReducerProps>  = {
  quotes: undefined,
  loading: false,
  error: null,
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const quoteReducer: any = (
  state = initialState,
  action: QuoteReducerProps
) => {
  switch (action.type) {
    case ADD_QUOTES:
      return typeof state === "object" && {
        ...state,
        quotes: action.quotes[0],
        loading: true,
        error: null,
      };
    case START_LOADING_DATA:
      return (
        typeof state === "object" && { ...state, loading: true }
      );
    case END_LOADING_DATA:
      return (
        typeof state === "object" && { ...state, loading: false }
      );
    case FETCH_ERROR:
      return { error: action.error }
    default:
      return state
  }
};

export default quoteReducer;
