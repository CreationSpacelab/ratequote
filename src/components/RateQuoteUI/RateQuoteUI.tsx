import React from "react";

import { useSelector, useDispatch } from "react-redux"

import { initiateAddQuotes } from "../../actions/quotes";
import { QuoteReducerProps, QuoteType, RateQuoteUIProps } from "../../quoteTypes";
import { ErrorMessage } from "../forms/FormStyles";
import { RateQuoteForm } from "../forms/RateQuoteForm";
import RateQuote from "../RateQuote";
import { LenderName, Column, TableHeaderRow } from "../RateQuote/RateQuoteStyles";

import { Container, Loader, Wrapper } from "./RateQuoteUIStyles";

export const DEFAULT_FORM_VALUE = {
  "loanSize": "450000",
  "creditScore": "680",
  "propertyType": "SingleFamily",
  "occupancy": "Primary"
}

const RateQuoteUI = () => {
  const dispatch = useDispatch()

  // Map state to props from redux using useSelector hook
  const quotes = useSelector((state: QuoteReducerProps) => state)

  const quotesPresent = quotes.quotes

  const handleSubmit = async (values: RateQuoteUIProps) => {
    dispatch(initiateAddQuotes(values))
  }

  // Sort quotes in descending order based on closingCosts
  const sortedQuotesDsc = quotes.quotes !== undefined &&
    quotes.quotes.sort((a, b) => a.closingCosts - b.closingCosts).reverse()

  return (
    <Wrapper>
      <RateQuoteForm setFormData={handleSubmit} />
      <Container>
        <TableHeader />
        {/* Map rate quotes to an array of RateQuote elements */}
        {quotesPresent ? !quotes.loading ? (
          <>
            {quotesPresent.length > 1 ? (
              Object.entries(sortedQuotesDsc).map(([k, quote]: [string, QuoteType]) => {
                return <RateQuote key={k} quote={quote} />
              })) : <ErrorMessageDisplay />}
          </>
        ) : (
            <Loader />
          ) :
          <ErrorMessage style={{ position: "absolute", left: "50%", transform: "translateX(-50%)"}}>
            <h3>Try a new search</h3>
          </ErrorMessage>
        }
      </Container>
    </Wrapper>
  );
};

/* wrapped in a memo() so it will only re-render when its props change. 
Each time the randomProp in RateQuote changes, mapStateToProps 
will be re-calculated */
export default React.memo(RateQuoteUI);

export const TableHeader: React.FC = () =>
  <TableHeaderRow>
    <LenderName data-testid="lenderName">Lender</LenderName>
    <Column data-testid="loanType">Product</Column>
    <Column data-testid="interestRateFormatted">Rate</Column>
    <Column data-testid="closingCosts">Closing <br /> Costs</Column>
    <Column data-testid="monthlyPaymentFormatted">Monthly Payment</Column>
    <Column data-testid="aprFormatted">APR</Column>
  </TableHeaderRow>

export const ErrorMessageDisplay: React.FC = () =>
  <ErrorMessage>
    <h3>Whoops!</h3>
    <h2>No results found.</h2>
    <p>Please try a different serach</p>
  </ErrorMessage>
