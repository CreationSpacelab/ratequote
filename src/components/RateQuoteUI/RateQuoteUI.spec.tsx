import React from "react";

import { cleanup, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux"
import thunk from "redux-thunk";

import { Action } from "../../actions/quotes";
import quoteReducer from "../../reducers/quotes";

import RateQuoteUI from "./RateQuoteUI";

// Create a HOC to render components with redux 
const renderWithRedux = (
  component: React.ReactNode,
  { store = createStore(quoteReducer) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  }
}

describe("RateQuoteUI", () => {
  afterEach(cleanup);
  
  test("renders app correctly", async () => {
    const { container } =
      renderWithRedux(<RateQuoteUI />
      );
    expect(container).toMatchSnapshot();
  });

  it("passes through an action", () => {
    const create = () => {
      const store = {
        getState: jest.fn(() => ({})),
        dispatch: jest.fn()
      }
      const next = jest.fn()
      const invoke = (action: Action) => thunk(store)(next)(action)

      return { store, next, invoke }
    }
    const { next, invoke } = create()
    const action: Action = { type: "START_LOADING_DATA", loading: true }
    invoke(action)
    expect(next).toHaveBeenCalledWith(action)
  });
})