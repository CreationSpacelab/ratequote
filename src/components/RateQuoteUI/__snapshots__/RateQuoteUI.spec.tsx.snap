// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateQuoteUI renders app correctly 1`] = `
<div>
  <div
    class="sc-jrAGrp jeFpNf"
  >
    <form
      class="sc-bdfBwQ cWDhuF"
    >
      <div>
        <div
          class="sc-gsTCUz ghpfRK"
        >
          <label
            for="loanSize"
          >
            Loan Size
          </label>
          <div
            class="sc-dlfnbm cATQrh"
          >
            <small>
              $
            </small>
            <input
              name="loanSize"
              placeholder="Enter loan size"
              type="text"
              value="450,000"
            />
          </div>
        </div>
        <div
          class="sc-gsTCUz OZGIW"
        >
          <label
            for="creditScore"
          >
            Credit Score
          </label>
          <div
            class="sc-dlfnbm cATQrh"
          >
            <input
              max="830"
              min="0"
              name="creditScore"
              pattern="[0-9]{3}"
              title="Must only contain numbers"
              type="number"
              value="680"
            />
          </div>
        </div>
      </div>
      <div>
        <div
          class="sc-gsTCUz OZGIW"
        >
          <label
            for="propertyType"
          >
            Property Type
          </label>
          <select
            name="propertyType"
          >
            <option
              value="SingleFamily"
            >
              Single Family
            </option>
            <option
              value="Condo"
            >
              Condo
            </option>
            <option
              value="Townhouse"
            >
              Town House
            </option>
            <option
              value="MultiFamily"
            >
              MultiFamily
            </option>
          </select>
        </div>
        <div
          class="sc-gsTCUz OZGIW"
        >
          <label
            for="occupancy"
          >
            Occupancy
          </label>
          <select
            name="occupancy"
          >
            <option
              value="Primary"
            >
              Primary Residence
            </option>
            <option
              value="Secondary"
            >
              Secondary
            </option>
            <option
              value="Investment"
            >
              Investment
            </option>
          </select>
        </div>
        <div
          class="sc-hKgILt iIJjpm"
        >
          <button
            data-testid="quote-rates-trigger"
            type="submit"
          >
            Quote Rates
          </button>
        </div>
      </div>
    </form>
    <div
      class="sc-kEjbxe bJgerX"
    >
      <div
        class="sc-iBPRYJ sc-fubCfw cVXiBE iRwpXQ"
      >
        <div
          class="sc-jSgupP sc-gKsewC bZDFWl cuROfH"
          data-testid="lenderName"
        >
          Lender
        </div>
        <div
          class="sc-jSgupP bZDFWl"
          data-testid="loanType"
        >
          Product
        </div>
        <div
          class="sc-jSgupP bZDFWl"
          data-testid="interestRateFormatted"
        >
          Rate
        </div>
        <div
          class="sc-jSgupP bZDFWl"
          data-testid="closingCosts"
        >
          Closing 
          <br />
           Costs
        </div>
        <div
          class="sc-jSgupP bZDFWl"
          data-testid="monthlyPaymentFormatted"
        >
          Monthly Payment
        </div>
        <div
          class="sc-jSgupP bZDFWl"
          data-testid="aprFormatted"
        >
          APR
        </div>
      </div>
      <div
        class="sc-eCssSg bOktpH"
        style="position: absolute; left: 50%; transform: translateX(-50%);"
      >
        <h3>
          Try a new search
        </h3>
      </div>
    </div>
  </div>
</div>
`;
